bpfman-operator/controllers/bpfman-agent/common.go

```diff
@@ -74,19 +74,234 @@ type ReconcilerCommon struct {
+		// updated and a reque has not been requested.
+		return ctrl.Result{Requeue: false}, nil
+	}
+}
+
+// reconcileBpfmanPrograms ONLY reconciles the bpfman state for a single BpfProgram.
+// It does not interact with the k8s API in any way.
+func (r *ReconcilerCommon) reconcileBpfmanProgram(ctx context.Context,
+	rec bpfmanReconciler,
+	existingBpfPrograms map[string]*gobpfman.ListResponse_ListResult,
+	bpfProgram *bpfmaniov1alpha1.BpfProgram,
+	isNodeSelected bool,
+	isBeingDeleted bool,
+	mapOwnerStatus *MapOwnerParamStatus) (bpfmaniov1alpha1.BpfProgramConditionType, error) {
+
+	r.Logger.V(1).Info("enter reconcileBpfmanProgram()", "bpfProgram", bpfProgram.Name, "CurrentProgram", rec.getName())
+
+	uuid := bpfProgram.UID
+
+	noContainers := noContainersOnNode(bpfProgram)
+
+	existingProgram, doesProgramExist := existingBpfPrograms[string(uuid)]
+	if !doesProgramExist {
+		r.Logger.V(1).Info("Program doesn't exist on node", "CurrentProgram", rec.getName())
+
+		// If Program is being deleted just exit
+		if isBeingDeleted {
+			return bpfmaniov1alpha1.BpfProgCondUnloaded, nil
+		}
+
+		// Make sure if we're not selected just exit
+		if !isNodeSelected {
+			return bpfmaniov1alpha1.BpfProgCondNotSelected, nil
+		}
+
+		// If a container selector is present but there were no matching
+		// containers on this node, just exit.
+		if noContainers {
+			r.Logger.V(1).Info("Program does not exist and there are no matching containers on this node")
+			return bpfmaniov1alpha1.BpfProgCondNoContainersOnNode, nil
+		}
+
+		// Make sure if the Map Owner is set but not found then just exit
+		if mapOwnerStatus.isSet && !mapOwnerStatus.isFound {
+			return bpfmaniov1alpha1.BpfProgCondMapOwnerNotFound, nil
+		}
+
+		// Make sure if the Map Owner is set but not loaded then just exit
+		if mapOwnerStatus.isSet && !mapOwnerStatus.isLoaded {
+			return bpfmaniov1alpha1.BpfProgCondMapOwnerNotLoaded, nil
+		}
+
+		// otherwise load it
+		loadRequest, err := rec.getLoadRequest(bpfProgram, mapOwnerStatus.mapOwnerId)
+		if err != nil {
+			return bpfmaniov1alpha1.BpfProgCondBytecodeSelectorError, err
+		}
+
+		r.Logger.Info("calling bpfman to load Program on Node", "bpfProgram Name", bpfProgram.Name)
+		r.progId, err = bpfmanagentinternal.LoadBpfmanProgram(ctx, r.BpfmanClient, loadRequest)
+		if err != nil {
+			r.Logger.Error(err, "Failed to load Program")
+			return bpfmaniov1alpha1.BpfProgCondNotLoaded, nil
+		}
+
+		return bpfmaniov1alpha1.BpfProgCondLoaded, nil
+	}
+
+	// prog ID should already have been set if program exists
+	id, err := bpfmanagentinternal.GetID(bpfProgram)
+	if err != nil {
+		r.Logger.Error(err, "Failed to get program ID")
+		return bpfmaniov1alpha1.BpfProgCondNotLoaded, nil
+	}
+
+	// BpfProgram exists but either Program is being deleted, node is no
+	// longer selected, or map is not available....unload program
+	if isBeingDeleted || !isNodeSelected || noContainers ||
```

Comment:
Should we break this up into multiple if statements for readability? It seems like we end up doing it anyways deeper in this if statement

Reply:
This is the way it was.  However, I'll see what I can do.

Reply:
I agree that this looks funky.  I think that the reason it was done this way is that for all these cases, the program needs to be unloaded, however, we return a different a different condition for each one.  The way it's written, we only have one call to UnloadBpfmanProgram().  If we break it up, we'd need to make the UnloadBpfmanProgram() call for each case.  Still looking for a way to make this whole function less confusing...

Reply:
I completely rewrote reconcileBpfProgram() into somethign that I think is more logical and readable.  Please let me know what you think.